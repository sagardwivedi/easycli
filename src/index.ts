import { createInterface } from "readline";
import chalk from "chalk";

export class Command {
  constructor(
    public name: string,
    public description: string,
    public callback: (args: string[], cli: CLI) => Promise<void> | void
  ) {}

  async execute(args: string[], cli: CLI): Promise<void> {
    await this.callback(args, cli);
  }
}

export class CLI {
  private commands: Map<string, Command>;

  constructor() {
    this.commands = new Map();
  }

  addCommand(
    name: string,
    description: string,
    callback: (args: string[], cli: CLI) => Promise<void> | void
  ): void {
    const lowerCaseName = name.toLowerCase();

    if (this.commands.has(lowerCaseName)) {
      console.error(
        chalk.red(
          `Command with name '${name}' already exists. Ignoring duplicate.`
        )
      );
      return;
    }

    const command = new Command(name, description, callback);
    this.commands.set(lowerCaseName, command);
  }

  showHelp(): void {
    console.log(chalk.bold("Available Commands:"));
    for (const [name, command] of this.commands) {
      console.log(`${chalk.green(name)}: ${command.description}`);
    }
  }

  async handleInput(input: string): Promise<void> {
    const trimmedInput = input.trim();

    if (trimmedInput === "") {
      console.log(
        chalk.yellow(
          'Please enter a valid command. Type "--help" for a list of available commands.'
        )
      );
      return;
    }

    const [commandName, ...args] = trimmedInput.split(" ");
    const lowerCaseCommandName = commandName.toLowerCase();
    const command = this.commands.get(lowerCaseCommandName);

    if (command) {
      try {
        await command.execute(args, this);
      } catch (error: any) {
        console.error(
          chalk.red(
            `Error executing command '${commandName}': ${
              error.message || error
            }`
          )
        );
      }
    } else if (lowerCaseCommandName === "--help") {
      this.showHelp();
    } else {
      console.log(
        chalk.red(
          `Unrecognized command '${commandName}'. Type "--help" for a list of available commands.`
        )
      );
    }
  }

  start(): void {
    const rl = createInterface({
      input: process.stdin,
      output: process.stdout,
      prompt: "> ",
    });

    rl.prompt();

    rl.on("line", async (line) => {
      const input = line.trim();

      try {
        await this.handleInput(input);
      } catch (error: any) {
        console.error(
          chalk.red(`Error processing input: ${error.message || error}`)
        );
      }

      rl.prompt();
    }).on("close", () => {
      console.log(chalk.yellow("Exiting CLI."));
      process.exit(0);
    });
  }
}
