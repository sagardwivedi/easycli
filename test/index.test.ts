import { expect, jest, test, spyOn } from "bun:test";
import { CLI } from "../src";

// Test command registration
test("addCommand prevents duplicate commands", async () => {
  const cli = new CLI();
  cli.addCommand("test", "Test command", async () => {});
  expect(() => cli.addCommand("test", "Another test", async () => {}))
    .toMatchSnapshot(`
      Error: Command with name 'test' already exists. Ignoring duplicate.
    `);
});

test("commands are case-insensitive", async () => {
  const cli = new CLI();
  cli.addCommand("test", "Test command", async () => {});
  expect(() => cli.addCommand("TEST", "Another test", async () => {}))
    .toMatchSnapshot(`
      Error: Command with name 'test' already exists. Ignoring duplicate.
    `);
});

// Test input handling
test("handles empty input", () => {
  const cli = new CLI();
  expect(async () => await cli.handleInput("")).toMatchSnapshot(
    'Please enter a valid command. Type "--help" for a list of available commands.'
  );
});

test("handles valid commands", async () => {
  const cli = new CLI();
  const mockCommand = jest.fn();
  cli.addCommand("validCommand", "Valid command", mockCommand);
  await cli.handleInput("validCommand");
  expect(mockCommand).toHaveBeenCalled();
});

test("handles unrecognized commands", () => {
  const cli = new CLI();
  expect(async () => await cli.handleInput("invalidCommand")).toMatchSnapshot(
    `Unrecognized command 'invalidCommand'. Type "--help" for a list of available commands.`
  );
});

// Test error handling
test("handles errors during command execution", async () => {
  const cli = new CLI();
  cli.addCommand("errorCommand", "Error command", async () => {
    throw new Error("Test error");
  });
  expect(async () => await cli.handleInput("errorCommand")).toMatchSnapshot(
    `Error executing command 'errorCommand': Test error`
  );
});

// Test start method
test("start listens for user input and handles commands", async () => {
  const cli = new CLI();
  const mockCommand = jest.fn();
  cli.addCommand("validCommand", "Valid command", mockCommand);

  const readline = require("readline").createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  spyOn(readline, "emit").mockImplementation((event, ...args) => {
    if (event === "line") {
      cli.handleInput(args[0]);
    }
  });

  cli.start();
  readline.emit("line", "validCommand");
  await Promise.resolve(); // Allow async operations to complete

  expect(mockCommand).toHaveBeenCalled();
});
