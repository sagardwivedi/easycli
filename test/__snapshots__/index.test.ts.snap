// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`addCommand prevents duplicate commands: 
      Error: Command with name 'test' already exists. Ignoring duplicate.
     1`] = `[Function]`;

exports[`commands are case-insensitive: 
      Error: Command with name 'test' already exists. Ignoring duplicate.
     1`] = `[Function]`;

exports[`handles empty input: Please enter a valid command. Type "--help" for a list of available commands. 1`] = `[Function: AsyncFunction]`;

exports[`handles unrecognized commands: Unrecognized command 'invalidCommand'. Type "--help" for a list of available commands. 1`] = `[Function: AsyncFunction]`;

exports[`handles errors during command execution: Error executing command 'errorCommand': Test error 1`] = `[Function: AsyncFunction]`;

exports[`showHelp displays available commands: 
  Available Commands:

  testCommand: Test command
   1`] = `[Function]`;
